tfworker/util/copier.py

```diff
@@ -0,0 +1,223 @@
+
+    @abstractstaticmethod
+    def type_match(source: str, **kwargs) -> bool:  # pragma: no cover
+        """ type_match determins if the source is supported/handled by a copier """
+        pass
+
+    @abstractmethod
+    def copy(self, **kwargs) -> bool:  # pragma: no cover
+        """ copy executes the copy from the source, into the working path """
+        pass
+
+    @property
+    def root_path(self):
+        """ root_path returns an optional root path to use for relative file operations """
+        if hasattr(self, "_root_path"):
+            return self._root_path
+        else:
+            return ""
+
+    @property
+    def conflicts(self):
+        """ conflicts returns a list of disallowed files """
+        if hasattr(self, "_conflicts"):
+            return self._conflicts
+        else:
+            return []
+
+    @property
+    def source(self):
+        """ source contains the source path providede """
+        return self._source
+
+    def get_destination(self, make_dir: bool = True, **kwargs) -> str:
+        """ get_destination returns the destination path, and optionally makes the destinatination directory """
+        if not (hasattr(self, "_destination") or "destination" in kwargs.keys()):
+            raise ValueError("no destination provided")
+        if "destination" in kwargs:
+            d = kwargs["destination"]
+        else:
+            d = self._destination
+
+        if make_dir:
+            make_d = Path(d)
+            make_d.mkdir(parents=True, exist_ok=True)
+
+        return d
+
+    def check_conflicts(self, path: str) -> None:
+        """ Checks for files with conflicting names in a path """
+        conflicting = []
+        if self.conflicts:
+            check_path = Path(path)
+            for check_file in check_path.glob("*"):
+                if check_file.name in self.conflicts:
+                    conflicting.append(check_file.name)
+
+        if conflicting:
+            raise FileExistsError(f"{','.join(conflicting)}")
+
+
+@CopyFactory.register("git")
+class GitCopier(Copier):
+    def copy(self, **kwargs) -> bool:
+        """ copy clones a remote git repo, and puts the requested files into the destination """
+        dest = self.get_destination(**kwargs)
+        branch = "master"
+
+        if "branch" in kwargs:
+            branch = kwargs["branch"]
+        if "sub_path" in kwargs:
+            sub_path = kwargs["sub_path"].strip("/")
+        else:
+            sub_path = ""
+
+        self.make_temp()
+        temp_path = f"{self._temp_dir}/{sub_path}"
+
+        pipe_exec(
+            f"git clone {self._source} --branch {branch} --single-branch ./",
+            cwd=self._temp_dir,
+        )
+
+        try:
+            self.check_conflicts(temp_path)
+        except FileExistsError as e:
+            self.clean_temp()
+            raise e
+
+        shutil.copytree(temp_path, dest, dirs_exist_ok=True)
+        self.clean_temp()
+
+        return True
```

Comment:
What does the returned bool mean? It looks like neither copier returns `False` ever and the calling site doesn't use the returned value.

Reply:
Good point, initially intended to return True/False if file copy was successful, but the function is not really setup (and doesn't need to be) to verify everything copied, and should rely on exceptions for any failures. Thanks!